tests/conftest.py

```diff
@@ -1,42 +1,139 @@
 # SPDX-License-Identifier: MIT
 
+import os
+import os.path
+import shutil
+import stat
+import sys
+import tarfile
+import tempfile
+
+import filelock
 import pytest
 
+import build
+
+
+if sys.version_info[0] == 2:
+    from urllib2 import urlopen
+else:
+    from urllib.request import urlopen
+
 
-DUMMY_PYPROJECT = '''
-[build-system]
-requires = ['flit_core']
-build-backend = 'flit_core.buildapi'
-'''.strip()
+INTEGRATION_SOURCES = {
+    'dateutil': ('dateutil/dateutil', '2.8.1'),
+    'pip': ('pypa/pip', '20.2.1'),
+    'Solaar': ('pwr-Solaar/Solaar', '1.0.3'),
+}
 
 
-DUMMY_PYPROJECT_NO_BACKEND = '''
-[build-system]
-requires = []
-'''.strip()
+def pytest_addoption(parser):
+    parser.addoption('--run-integration', action='store_true', help='run the integration tests')
+    parser.addoption('--only-integration', action='store_true', help='only run the integration tests')
 
 
-DUMMY_PYPROJECT_NO_REQUIRES = '''
-[build-system]
-build-backend = 'something'
-'''.strip()
+def pytest_collection_modifyitems(config, items):
+    skip_int = pytest.mark.skip(reason='integration tests not run')
+    skip_other = pytest.mark.skip(reason='only integration tests are run')
+
+    for item in items:
+        if os.path.basename(item.location[0]) == 'test_integration.py':
+            if not config.getoption('--run-integration') and not config.getoption('--only-integration'):
+                item.add_marker(skip_int)
+        elif config.getoption('--only-integration'):
+            item.add_marker(skip_other)
 
 
 @pytest.fixture
-def empty_file_mock(mocker):
-    return mocker.mock_open(read_data='')
+def packages_path():
+    return os.path.realpath(os.path.join(__file__, '..', 'packages'))
+
+
+@pytest.fixture(scope='session')
+def integration_path():
+    srcdir = os.path.realpath('.integration-sources')
+    dest = tempfile.mkdtemp(prefix='python-build-integration-')
+
+    if not os.path.exists(srcdir):
+        os.makedirs(srcdir)
+
+    # self
+    self_source = os.path.abspath(os.path.join(__file__, '..', '..'))
+    self_dest = os.path.join(dest, 'python-build')
+    if not os.path.exists(self_dest):
+        if build.env.fs_supports_symlink():
+            os.symlink(self_source, self_dest)
+        else:  # pragma: no cover
+            shutil.copytree(self_source, self_dest)
+
+    for target, (repo, version) in INTEGRATION_SOURCES.items():
+        with filelock.FileLock(os.path.join(srcdir, '{}.lock'.format(target))):
+            target_dest = os.path.join(dest, target)
+            if os.path.exists(target_dest):  # pragma: no cover
+                continue
+
+            tarball = os.path.join(srcdir, '{}.tar.gz'.format(target))
+            data = urlopen('https://github.com/{}/archive/{}.tar.gz'.format(repo, version)).read()
+            with open(tarball, 'wb') as f:
```

Comment:
instead of ``f`` use ``file_handler`` one character variables should be used only for list comprehension variables IMHO

Reply:
I disagree in this specific case, `f` is a pretty standard name for file context managers. And here we don't have any other file variable so it won't cause any confusion. But I can change if you want.

Reply:
I personally agree on this with PEP-8 and prefer it, but 🤷